File: longestSubsequence.js
function longestSubsequence(numbers) {
	var result = 0;
	numbers.reduce(function(pv, cv) {
		if (pv.length == 0) {
			result = 1;
			return [{value: cv, maxForThis: 1}];
		}
		var max = 1;
		pv.forEach(function(each) {
			if (each.value < cv) {
				if (each.maxForThis + 1 > max) max = each.maxForThis + 1;
			}
		});
		pv.push({value: cv, maxForThis: max});
		if (result < max) result = max;
		return pv;
	}, []);
	return result;
}

console.log(longestSubsequence([0, 8, 4, 12, 2, 10, 6, 14, 1, 9, 5, 13, 3, 11, 7, 15]))

File: longestSubsequence.py
def longestSubsequence(numbers):
	maxArray = [i for i in range(len(numbers))]
	for i, value in enumerate(numbers):
		maxArray[0] = 1
		maxValue = 1
		for j in range(0, i):
			if numbers[j] < value and maxValue < maxArray[j] + 1:
				maxValue = maxArray[j] + 1
		maxArray[i] = maxValue
		print(maxArray)
	return 0 if len(maxArray) == 0 else max(maxArray);

for x in range(int(input())):
	numberCount = input()
	numbers = [ int(i) for i in input().split() ]
	print(longestSubsequence(numbers))
